### ISSUE TITLE
Ardour does not save the parameters in lv2 plugins exported by Carla

### ISSUE THREAD

**Issue by RockVice77** [falkTX/Carla#831]
Hello,
I am still experimenting with several winVST plugins exported by Carla as LV2: my current issue is that any exported LV2 plugin loaded in Ardour restores the default parameters every time I save and reopen the session
The issue is not present if I save an Ardour session with the same plugin loaded in a Carla-rack LV2: the parameters of the plugin are properly saved and they are applied when I re-open the session

I am not able to say if this is a bug on Carla or Ardour side

**Comment by manhnt9** [comment]
I can confirm that this bug presents. I'm testing with Native Instrument's Kontakt Player.
Carla reloads Kontakt's instruments when reloading a saved Carla project.
But with LV2 exporting, Ardour can't remember.
Not to mention that there're many Ardour crashes when using Kontakt as an exported LV2 plugin.

**Comment by jerboyd** [comment]
I am using Carla 2.0.90 and Ardour 5/Mixbus 5 and I also have this behaviour. the exported plugins load properly and work quite well with no hang-ups so far but plugin parameters are never saved. Has there been any progress on this?

**Comment by falkTX** [comment]
This is not implemented yet, so of course it won't work.
The LV2 export is experimental because it is not finished. Sorry it was not clear before.
I started it as a test to see how well it behaved, if it could even work even. But seeing as it seems to work well, it is best to just go and finish it.

**Comment by Floby** [comment]
Hello !
I too tried to load an exported lv2 plugin (bridging to wine). I only managed to load it for the vst2 version of the plugin (not vst3) and couldn't either control or automate from ardour
This is impressive work ! Thank you =)

I like to format my feature requests as pull requests ^^. Reading around, I found that most of this functionality seems to be implemented [here](https://github.com/falkTX/Carla/blob/master/source/bridges-plugin/CarlaBridgeSingleLV2.cpp) for the runtime. and also [here](https://github.com/falkTX/Carla/blob/c14ec23ab96adab14ae298a5bdb7c5d065e9aa0f/source/backend/plugin/CarlaPlugin.cpp#L1010) for the bundle generation.

However, I don't understand why these input controls are not working as expected as they seem to be properly connected [here](https://github.com/falkTX/Carla/blob/master/source/bridges-plugin/CarlaBridgeSingleLV2.cpp#L484)

Can you tell me what I'm missing ?

**Comment by falkTX** [comment]
You are quite correct on the code place! Nice that you figured that out by yourself.

Now... the parameters are mapped actually, but the way VST plugins work is that they mostly ignore those parameters during save/load. Most VST plugins use a "chunk" for saving their state, which includes parameter values. Carla will not restore parameter values if the plugin uses these "chunks".

One quick workaround is to disable the use of "Chunks" before exporting to LV2. So that Carla will take care of parameter save&load. But if the plugin has any non-parameter data (text field, drawing canvas, etc) those will be ignored.

The way to finish this feature is:
1. export state save&restore to the carla-lv2-export plugin
2. save&restore plugin "chunks" during those state calls
3. (only needed for DSSI and LV2) map the custom data into lv2 state

**Comment by Floby** [comment]
Thank you very much for your reply. I consumed a lot of my c++ deciphering budget already so I will look at it tomorrow =)
I shall give it a try but there's no guarantee, because it's been a while since I coded in C++.
